README


This archive contains the distribution CPAN-Meta-Check,
version 0.009:

  Verify requirements in a CPAN::Meta object

This software is copyright (c) 2012 by Leon Timmermans.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.019.



Changes
Revision history for CPAN-Meta-Check

0.009     2014-06-20 19:37:43+02:00 Europe/Amsterdam
          Various POD fixes

0.008     2013-10-17 16:44:57 Europe/Amsterdam
          Switch to using merged_requirements
          Test Env instead of Carp for version overshoot [#89591]
          Document $incdirs in the right function

0.007     2013-07-31 14:37:58 Europe/Amsterdam
          Swap conflicts test, as underscore versions broke it

0.006     2013-07-27 00:24:21 Europe/Amsterdam
          Fixed bad dereference during conflicts checking

0.005     2013-04-22 17:27:50 Europe/Amsterdam
          Converted from Test::Differences to Test::Deep
          Don't load CPAN::Meta from CPAN::Meta::Check

0.004     2012-05-24 21:35:08 Europe/Amsterdam
          Depend on recent enough CPAN::Meta

0.003     2012-05-09 20:48:01 Europe/Amsterdam
          Allow explicit setting of the include dir

0.002     2012-04-05 01:05:34 Europe/Amsterdam
          Add explicit dependency on CPAN::Meta 2.120920
          Allow CPAN::Meta::Requirements instead of CPAN::Meta

0.001     2012-04-01 12:46:04 Europe/Amsterdam
          Initial release


lib/CPAN/Meta/Check.pm
package CPAN::Meta::Check;
{
  $CPAN::Meta::Check::VERSION = '0.006';
}
use strict;
use warnings;

use Exporter 5.57 'import';
our @EXPORT = qw//;
our @EXPORT_OK = qw/check_requirements requirements_for verify_dependencies/;
our %EXPORT_TAGS = (all => [ @EXPORT, @EXPORT_OK ] );

use CPAN::Meta::Requirements 2.120920;
use Module::Metadata;

sub _check_dep {
	my ($reqs, $module, $dirs) = @_;

	my $version = $module eq 'perl' ? $] : do { 
		my $metadata = Module::Metadata->new_from_module($module, inc => $dirs);
		return "Module '$module' is not installed" if not defined $metadata;
		eval { $metadata->version };
	};
	return "Missing version info for module '$module'" if $reqs->requirements_for_module($module) and not $version;
	return sprintf 'Installed version (%s) of %s is not in range \'%s\'', $version, $module, $reqs->requirements_for_module($module) if not $reqs->accepts_module($module, $version || 0);
	return;
}

sub _check_conflict {
	my ($reqs, $module, $dirs) = @_;
	my $metadata = Module::Metadata->new_from_module($module, inc => $dirs);
	return if not defined $metadata;
	my $version = eval { $metadata->version };
	return "Missing version info for module '$module'" if not $version;
	return sprintf 'Installed version (%s) of %s is in range \'%s\'', $version, $module, $reqs->requirements_for_module($module) if $reqs->accepts_module($module, $version);
	return;
}

sub requirements_for {
	my ($meta, $phases, $type) = @_;
	my $prereqs = ref($meta) eq 'CPAN::Meta' ? $meta->effective_prereqs : $meta;
	if (!ref $phases) {
		return $prereqs->requirements_for($phases, $type);
	}
	else {
		my $ret = CPAN::Meta::Requirements->new;
		for my $phase (@{ $phases }) {
			$ret->add_requirements($prereqs->requirements_for($phase, $type));
		}
		return $ret;
	}
}

sub check_requirements {
	my ($reqs, $type, $dirs) = @_;

	my %ret;
	if ($type ne 'conflicts') {
		for my $module ($reqs->required_modules) {
			$ret{$module} = _check_dep($reqs, $module, $dirs);
		}
	}
	else {
		for my $module ($reqs->required_modules) {
			$ret{$module} = _check_conflict($reqs, $module, $dirs);
		}
	}
	return \%ret;
}

sub verify_dependencies {
	my ($meta, $phases, $type, $dirs) = @_;
	my $reqs = requirements_for($meta, $phases, $type);
	my $issues = check_requirements($reqs, $type, $dirs);
	return grep { defined } values %{ $issues };
}

# vi:noet:sts=2:sw=2:ts=2
1;

#ABSTRACT: Verify requirements in a CPAN::Meta object

__END__

=pod

=head1 NAME

CPAN::Meta::Check - Verify requirements in a CPAN::Meta object

=head1 VERSION

version 0.006

=head1 SYNOPSIS

 warn "$_\n" for verify_requirements($meta, [qw/runtime build test/], 'requires');

=head1 DESCRIPTION

This module verifies if requirements described in a CPAN::Meta object are present.

=head1 FUNCTIONS

=head2 check_requirements($reqs, $type)

This function checks if all dependencies in C<$reqs> (a L<CPAN::Meta::Requirements|CPAN::Meta::Requirements> object) are met, taking into account that 'conflicts' dependencies have to be checked in reverse. It returns a hash with the modules as keys and any problems as values; the value for a successfully found module will be undef.

=head2 verify_dependencies($meta, $phases, $types, $incdirs)

Check all requirements in C<$meta> for phases C<$phases> and types C<$types>. Modules are searched for in C<@$incdirs>, defaulting to C<@INC>.

=head2 requirements_for($meta, $phases, $types, incdirs)

This function returns a unified L<CPAN::Meta::Requirements|CPAN::Meta::Requirements> object for all C<$type> requirements for C<$phases>. $phases may be either one (scalar) value or an arrayref of valid values as defined by the L<CPAN::Meta spec|CPAN::Meta::Spec>. C<$type> must be a relationship as defined by the same spec. Modules are searched for in C<@$incdirs>, defaulting to C<@INC>.

=head1 SEE ALSO

=over 4

=item * L<Test::CheckDeps|Test::CheckDeps>

=item * L<CPAN::Meta|CPAN::Meta>

=back

=head1 AUTHOR

Leon Timmermans <leont@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Leon Timmermans.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Verify requirements in a CPAN::Meta object",
  "AUTHOR" => "Leon Timmermans <leont\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "CPAN-Meta-Check",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "CPAN::Meta::Check",
  "PREREQ_PM" => {
    "CPAN::Meta::Prereqs" => "2.132830",
    "CPAN::Meta::Requirements" => "2.121",
    "Exporter" => "5.57",
    "Module::Metadata" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "CPAN::Meta" => "2.120920",
    "Env" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Deep" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.009",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::Meta" => "2.120920",
  "CPAN::Meta::Prereqs" => "2.132830",
  "CPAN::Meta::Requirements" => "2.121",
  "Env" => 0,
  "Exporter" => "5.57",
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Metadata" => 0,
  "Test::Deep" => 0,
  "Test::More" => "0.88",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



